[EuiDataGrid] use schema information when sorting

chandlerprall opened the pull request:
### Summary

Resolves #2399

Uses schema information (both provided and detected) for better in-memory column sorting. The gif below shows the `amount` column sorted by its `currency` comparator instead of lexicographically.

![in-memory schema-based sorting](https://d.pr/i/H0E1aX.gif)

myasonik commented:
Users can pass in a supported schema, right? But can users pass in a custom schema/comparator?

chandlerprall commented:
> Users can pass in a supported schema, right?

Supported or even a custom type, which leads to...

> But can users pass in a custom schema/comparator?

Yes! They can define a schema detector and comparator for a custom type, or even override an type's functionality.

snide commented:
Adding json schema which seems to work. also found an issue with empty popovers that i fixed. unfortunately, this breaks a test, because it's expecting the popover to be there in certain cases.

**This was the bug**

![image](https://user-images.githubusercontent.com/324519/66671800-af3b2980-ec11-11e9-9084-1a843efb5b13.png)

**Resolved in the following scenarios now**

![image](https://user-images.githubusercontent.com/324519/66671761-97fc3c00-ec11-11e9-86ee-78e20450b76c.png)

![image](https://user-images.githubusercontent.com/324519/66671778-a34f6780-ec11-11e9-9f2a-b6ea04630bd0.png)

chandlerprall commented:
Fixed the unit test and updated currency detector to be less aggressive.

> I did notice that the numeric and currency comparator functions are identical.

By design, as they both want to extract & sort on the same underlying numerical data. But good eyes and thanks for verifying!
